include/zephyr/drivers/battery.h

```diff
@@ -0,0 +1,470 @@
+	int desired_current; /* Charging current desired by battery (mA) */
+	int remaining_capacity; /* Remaining capacity in mAh */
+	int full_capacity; /* Capacity in mAh (might change occasionally) */
+	int display_charge; /* Display charge in 10ths of a % (1000=100.0%) */
+	int status; /* Battery status */
+	enum battery_present is_present; /* Is the battery physically present */
+	int flags; /* Flags */
+};
+
+struct ship_mode_info {
+	/*
+	 * Write Block Support. If wb_support is true, then we use a i2c write
+	 * block command instead of a 16-bit write. The effective difference is
+	 * that the i2c transaction will prefix the length (2) when wb_support
+	 * is enabled.
+	 */
+	const uint8_t wb_support;
+	const uint8_t reg_addr;
+	const uint16_t reg_data[CONFIG_BATTERY_SHIP_MODE_WRITES];
+};
+
+/* Battery constants */
+struct battery_info {
+	/* Operation voltage in mV */
+	int voltage_max;
+	int voltage_normal;
+	int voltage_min;
+	/**
+	 * Pre-charge to fast charge threshold in mV,
+	 * default to voltage_min if not specified.
+	 * This option is only available on isl923x and rt946x.
+	 */
+	int precharge_voltage;
+	/* Pre-charge current in mA */
+	int precharge_current;
+	/* Working temperature ranges in degrees C */
+	int8_t start_charging_min_c;
+	int8_t start_charging_max_c;
+	int8_t charging_min_c;
+	int8_t charging_max_c;
+	int8_t discharging_min_c;
+	int8_t discharging_max_c;
+	uint8_t vendor_param_start;
+};
+
+/*
+ * Provide a 1 minute average of the current and voltage on the battery.
+ * Does not check for flags or whether those values are bad readings.
+ * See driver/battery/[your_driver].h/c for details on implementation and
+ * how the average is calculated.
+ */
+typedef int (*battery_get_avg_current)(const struct device *dev); /* in mA */
+typedef int (*battery_get_avg_voltage)(const struct device *dev); /* in mV */
+
+/**
+ * Return vendor-provided battery constants.
+ */
+typedef const struct battery_info *(*battery_get_info)(const struct device *dev);
+
+/**
+ * Get current battery parameters.
+ *
+ * Error conditions are reported via batt.flags.
+ *
+ * @param batt		Destination for battery data
+ */
+typedef void (*battery_get_params)(const struct device *dev, struct batt_params *batt);
+
+/**
+ * Modify battery parameters to match vendor charging profile.
+ *
+ * @param batt		Battery parameters to modify
+ */
+typedef void (*battery_override_params)(const struct device *dev, struct batt_params *batt);
+
+/**
+ * Check for presence of battery.
+ *
+ * @return Whether there is a battery attached or not, or if we can't tell.
+ */
+typedef enum battery_present (*battery_is_present)(const struct device *dev);
+
+/**
+ * Check for physical presence of battery.
+ *
+ * @return Whether there is a battery physically present, but possibly
+ * in a disconnected or cut off state, or if we can't tell;
+ */
+typedef enum battery_present (*battery_hw_present)(const struct device *dev);
```

Comment:
I think battery_hw_present and battery_is_present should be combined into one, returning an enum (operating, switched_off,missing,unknown), partly because I'm not keen on the name battery_hw_present!

Is there a need to record the state 'operating but not chargeable', e.g. for an alkaline battery or where charging has been tried but fails?